src/app/reporting/Engine.cpp

```diff
@@ -336,6 +377,78 @@ void Engine::Run()
         mCurReadHandlerIdx = (mCurReadHandlerIdx + 1) % CHIP_IM_MAX_NUM_READ_HANDLER;
         readHandler        = imEngine->mReadHandlers + mCurReadHandlerIdx;
     }
+    for (auto & path : mDirtyPathList)
+    {
+        path.ClearDirty();
+    }
+    return;
+}
+
+void Engine::SetDirty(ClusterInfo & aClusterInfo)
+{
+    InteractionModelEngine * imEngine = InteractionModelEngine::GetInstance();
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
+    ScopedLock lock(*this);
+#endif // !CHIP_SYSTEM_CONFIG_NO_LOCKING
+
+    for (auto & path : mDirtyPathList)
+    {
+        if (path.IsDirty())
+        {
+            //If aClusterInfo is part of existing dirty path, skip it
+            //If aClusterInfo is super set of one existing path, update that path with new path.
+            if (path.IsAttributePathSupersetOf(aClusterInfo)) {
+                return;
+            }
+            if (aClusterInfo.IsAttributePathSupersetOf(path)) {
+                path.mListIndex = aClusterInfo.mListIndex;
+                path.mFieldId = aClusterInfo.mFieldId;
+                path.mFlags = aClusterInfo.mFlags;
+                return;
+            }
+        }
+    }
+
+    for (auto & readHandler : imEngine->mReadHandlers)
+    {
+        if (readHandler.IsSubscription() && readHandler.IsReportable())
+        {
+            ClusterInfo * clusterInstance = readHandler.GetAttributeClusterInfolist();
+
+            while (clusterInstance != nullptr)
+            {
+                bool setSuccess = false;
+                // If dirty path intersects with interested path from handler, it mark this in dirty path list and also
+                // mark cluster as dirty for the particular subscribe handler , otherwise, it skip this dirty path.
+                if (clusterInstance->IsAttributePathSupersetOf(aClusterInfo) ||
+                    aClusterInfo.IsAttributePathSupersetOf(*clusterInstance))
+                {
+                    for (auto & path : mDirtyPathList)
```

Comment:
Can we move this outside the very outter loop?
i.e.
1. Get a seat in dirty path list
2. Iterate all client and mark necessary path dirty.
3. If we marked any path dirty, set that item dirty.

Reply:
the intuitive loop
-- Loop all handle, 
-- then loop corresponding interested clusters for each handler
--then check it against the dirty path list
it seems it is more suitable here.